projects/components/src/table/renderers/data-cell/table-data-cell-renderer.component.ts

```diff
@@ -1,54 +1,95 @@
-import { ChangeDetectionStrategy, Component, Injector, Input, OnInit } from '@angular/core';
-import { Dictionary } from '@hypertrace/common';
-import { TableColumnConfig } from '../../table-api';
+import {
+  ChangeDetectionStrategy,
+  Component,
+  ComponentFactoryResolver,
+  ComponentRef,
+  Injector,
+  Input,
+  OnInit
+} from '@angular/core';
+import { FilterAttribute } from '../../../filter-bar/filter-attribute';
+import { TableColumnConfig, TableRow } from '../../table-api';
 import { TableCellAlignmentType } from '../table-cell-alignment-type';
 import { TableCellRendererConstructor } from '../table-cell-renderer';
+import { TableCellRendererComponent } from '../table-cell-renderer.component';
 import { TableCellRendererService } from '../table-cell-renderer.service';
 
 @Component({
   selector: 'htc-table-data-cell-renderer',
   styleUrls: ['./table-data-cell-renderer.component.scss'],
   changeDetection: ChangeDetectionStrategy.OnPush,
   template: `
-    <div class="vertical-alignment-container">
-      <div class="horizontal-alignment-container" [ngClass]="this.alignment">
+    <div class="table-data-cell-renderer" [class.selected]="this.popoverOpen">
+      <htc-filter-button
+        class="filter-button"
+        *ngIf="this.columnConfig?.filterable && this.leftAlignFilterButton"
+        [metadata]="this.metadata"
+        [attribute]="this.columnConfig?.attribute"
+        [value]="this.componentRef?.instance.value"
```

Comment:
what can we do to avoid reaching into a component from its parent

Reply:
The issues you noted in this file are the ugly parts I alluded to. Agree we need a better way. Let's have the conversation here and I'll link the other comments about the cell renderer code to this one.

Reply:
so my starting question here would actually be - do we need the renderer at all to get the value? I would think filtering generally should be based off the raw value rather than however the renderer packages it for display - although this can actually go both ways so neither way is perfectly correct. In one direction for example, take the backend icon renderer - it takes a backend type and converts it into an icon type for rendering. If we were to filter on that, we'd want to take that raw backend type. In the other direction, something like metric currently gets mapped to an object containing units, then unpacked in the renderer.

So, I guess what I'm trying to say is that we've overloaded the concept of value in renderer - the value to use for building a filter often overlaps with the raw value coming in, or the renderer produced value, but both are something different. Given that, the most appropriate place (ignoring implementation) I would think would be in the specification - that's where we define `asGraphQlOrderByFragment` which is a similar use case.

Reply:
So there's 3 actions we do with a a column/cell right now -
1. Display it. This takes in a value, and parses it to some type of visualization. It could do things like adding units, changing formats, joining extra data etc. Nothing here cares about where the data came from.
2. Sort it. This takes a column definition and converts it into a sort command. The sort command itself depends on the implementation of the data source, so we tell the data source which column was sorted as well as the direction,  and it handles how to do so (`TableDataRequest`). In converting a column to a sort, the specific implementation of column is responsible for determining how to materialize that. In our case, it's a graphql specification which could map to multiple attributes because we have some complex columns. Take status for example, where we join status the boolean, status code and status description. When we sort it, we want to just use the single status code column.
3. Now, filter it. This is almost a combination of 1+2 since it needs both an attribute and value, but IMO it's much closer to 2 because the value we need here needs to match the data source expected value, not the display value. The graphql specification was responsible for combining any number of attributes and producing a value out of them, so already has the knowledge and seems the appropriate place to do that in reverse.

Reply:
So, what is the ask here then?

Reply:
I'm not sure. I'd rather have the spec deal with this, but given the level of effort, I think the only realistic ask is to clean up how we interact with the renderer. So my compromise ask is to not create two copies of every renderer but use the one actual rendering to get the value, rather than creating and discarding a new component.

Reply:
That's reasonable. Let me take a look on getting that working.